Start seriously converting the CLI to TypeScript

pimterry opened the pull request:
This PR converts almost all the utils, a little of the easiest commands & all the documentation generation pipeline to TypeScript. ~~This depends on https://github.com/resin-io/resin-sdk/pull/448 and https://github.com/resin-io-modules/stream-logger/pull/4, so won't build until those are complete (though it does build locally for me, all linked together).~~

I've skipped the Docker utils (since I think @dfunckt is creating a lot of conflicting changes in those as we speak), and I haven't started on the top-level actions etc for now, because that'll be both very conflicty, fairly tricky, this would be hard to review all in one go!

This is still pretty big though - might be a little easier to review commit by commit.

I've tried to be pretty cautious sticking with existing behaviour here for now, and just directly translating, to keep this manageable. That said, there's a two notable discrepancies that I've fixed where TS found real bugs (which is cool):

* Our current doc generation rendered options with multiple aliases incorrectly, e.g. the `--email`/`-e`/`-u` argument to resin login: https://github.com/resin-io/resin-cli/blob/master/doc/cli.markdown#login.
* `getManifest` in helpers.coffee was falling back to the API 100% of the time, because it assumed rindle had already been promisified, tried to call `extractAsync` (which doesn't exist), and then silently checked the API if it failed.

There's also a fun new TS pattern I'm using here: [dynamic import expressions](https://github.com/basarat/typescript-book/blob/master/docs/project/dynamic-import-expressions.md). Lots of CLI code runs `require` inside functions (to lazily load modules), which TS (and ES6 modules) don't allow.

Instead, we use `import()` to do this async, plus async functions to make that nice and readable, and a couple of cases where we have to import the types only seperately outside the function to make it work. Nicest solution I can find, but suggestions welcome.

thgreasi commented on lib/utils/streams.ts:
```diff
@@ -0,0 +1,20 @@
+		.pipe(fileWriteStream);
+	}).then(() => new Promise(function(resolve, reject) {
+		fs.createReadStream(bufferFile)
+		.on('open', function(this: ReadStream) {
```
Could this be `.on('open', resolve)` and use `new Promise<ReadStream>`?
Actually I just tried that and got `Promise<ReadStream>` instead of `Promise<{}>` inferred as the return type of the function.

pimterry replied:
> Could this be .on('open', resolve)

Nope, I think the `this` argument isn't doing what you think here. That doesn't mean that there's a first argument called `this`. It's just telling typescript the implicit `this` that the function will be called with. This code is equivalent to:

```js
.on('open', function () {
  resolve(this);
})
```

It's a bit tricky: we're basically saying 'when the stream opens, resolve with the stream itself', without having an actual stream variable flying around anywhere. You're totally right though, it was confusingly written, I've simplified it.

thgreasi replied:
Ohhhh, :sparkles: :sparkling_heart: :heart_eyes:

thgreasi replied:
Is there a reason that the `ReadStream` isn't used any more? The `buffer()` now returns `Promise<{}>`. Should we use `new Promise<ReadStream>` or similar casting to fix that?

thgreasi commented on lib/utils/patterns.ts:
```diff
@@ -0,0 +1,228 @@
+const resin = ResinSdk.fromSharedOptions();
+
+export function authenticate(options: {}) {
+	form.run([{
```
Is it correct fine that we removed the `return`? Since, in [lib/actions/auth.coffee](https://github.com/resin-io/resin-cli/blob/810ca78215ee1dd71e0dd84506476e5af71da8be/lib/actions/auth.coffee#L92) it seems that we return the result of this function.

pimterry replied:
Fixed, great catch! That would've broken manually entered logins, nice spot.

thgreasi commented on lib/utils/patterns.ts:
```diff
@@ -0,0 +1,228 @@
+		return form.ask({
+			message: 'Two factor auth challenge:',
+			name: 'code',
+			type: 'input'}).then(resin.auth.twoFactor.challenge)
```
This will be fixed by prettier, but I would love to see the `.then` in the next line right away :)

pimterry replied:
Fixed

thgreasi commented on lib/utils/patterns.ts:
```diff
@@ -0,0 +1,228 @@
+		if (!hasAnyApplications) return;
+
+		return resin.models.application.getAll().then((applications) => {
+			let appOptions: { name: string, value: string | null }[];
```
NIT: This could be a `const` with the object type defined as a generic parameter of `map`.

pimterry replied:
Fixed (the generics end up being a bit more awkward than you'd imagine, and inference can't handle it as we need to add a `null` value later, but the end result's not too bad)

thgreasi commented on lib/utils/patterns.ts:
```diff
@@ -0,0 +1,228 @@
+export function inferOrSelectDevice(preferredUuid: string) {
+	return resin.models.device.getAll()
+	.filter((device: ResinSdk.Device) => device.is_online)
+	.then((onlineDevices: ResinSdk.Device[]) => {
```
NIT: This could be changed to
```ts
.filter<ResinSdk.Device>(device => device.is_online)
.then(onlineDevices) => {
```
and the `.then` would properly infer the type.

pimterry replied:
Fixed

thgreasi commented on lib/utils/patterns.ts:
```diff
@@ -0,0 +1,228 @@
+			throw new Error('You don\'t have any devices online');
+		}
+
+		let defaultUuid = Array.from(_.map(onlineDevices, 'uuid')).includes(preferredUuid) ?
```
Also I don't think that we need `Array.from` but I guess that this is just decaffeinate.

pimterry replied:
Both true, fixed

thgreasi commented on lib/utils/config.ts:
```diff
@@ -0,0 +1,109 @@
+		registryUrl: resin.settings.get('registryUrl'),
+		deltaUrl: resin.settings.get('deltaUrl'),
+		pubNubKeys: resin.models.config.getAll().get('pubnub'),
+		mixpanelToken: resin.models.config.getAll().get('mixpanelToken')
```
Could we call `getAll()` only once? It seems to send separate requests.

pimterry replied:
True! Good catch, that would've been annoying - now fixed.

thgreasi commented on lib/utils/logger.ts:
```diff
@@ -0,0 +1,66 @@
+			error: logger.createLogStream('error')
+		};
+
+		_.mapKeys(this.streams, function(stream, key) {
```
This should probably be `forEach` since we aren't doing anything with its results and also some `return`s could be removed as a result.

pimterry commented on extras/capitanodoc/index.ts:
```diff
@@ -0,0 +1,33 @@
+	for (let file of commandCategory.files) {
+		const actions: any = require(path.join(process.cwd(), file));
+
+		if (actions.signature != null) {
```
Nope, sensible, done.

pimterry commented on extras/capitanodoc/index.ts:
```diff
@@ -0,0 +1,33 @@
+		if (actions.signature != null) {
+			category.commands.push(_.omit(actions, 'action'));
+		} else {
+			for (let actionName in actions) {
```
Strangely it didn't seem to mind, maybe because `actions` is `any`? No idea, but I've moved to `for of Object.keys` instead, which is better regardless.

pimterry commented on extras/capitanodoc/markdown.ts:
```diff
@@ -0,0 +1,90 @@
+
+export function renderCommand(command: Command) {
+	let result = `\
+## ${ent.encode(command.signature)}
```
It is weird. I played with common tags briefly but it seemed to get confused by some of the indentation here. Not sure why, but I've just simplified these to single line strings and explicit newlines instead, since they're only short.

pimterry commented on extras/capitanodoc/markdown.ts:
```diff
@@ -0,0 +1,90 @@
+
+export function renderCategory(category: Category) {
+	let result = `\
+# ${category.title}\n\
```
Fixed

pimterry commented on extras/capitanodoc/utils.ts:
```diff
@@ -0,0 +1,35 @@
+		if (_.isString(option.alias)) {
+			result += `, ${getOptionSignature(option.alias)}`;
+		} else {
+			for (let alias of option.alias!) {
```
I'm not sure what this means. Which `if` - the `isString`, or `isEmpty`? I can't see how we can easily change either one to fix this... I could change the `else` to an explicit `else if` and check there, but that changes what this code actually does slightly.

Really, TS should know this can't be undefined because of the `isEmpty`, but the lodash typings don't include a type guard for that (and I think it's impossible to write a 100% correct one, sadly).

thgreasi replied:
Opps, sorry for being confusing... I was meaning something like:
```ts
if (_.isArray(option.alias)) {
	for (let alias of option.alias) { /*...*/ }
} else {
	result += `, ${getOptionSignature(option.alias)}`;
}
```
but it's an NIT anyway.

pimterry replied:
Ah, got it. Fixed!

pimterry commented on extras/capitanodoc/utils.ts:
```diff
@@ -0,0 +1,35 @@
+		}
+	}
+
+	if (option.parameter != null) {
```
Done

pimterry commented on extras/tsconfig.json:
```diff
@@ -0,0 +1,17 @@
+		"noUnusedParameters": true,
+		"preserveConstEnums": true,
+		"removeComments": true,
+		"sourceMap": true
```
`strict: true` includes both the first two (and a few other nice settings too), and `node` resolution is the default with `"module": "commonjs"` anyway.

thgreasi replied:
Didn't know that part WRT "module": "commonjs" !

pimterry commented on lib/actions/settings.ts:
```diff
@@ -12,23 +12,27 @@ distributed under the License is distributed on an "AS IS" BASIS,
+	$ resin settings\
+`,
+	async action(_params, _options, done) {
+		const resin = (await import('resin-sdk')).fromSharedOptions();
```
Yep, it doesn't make any direct difference in this one case, I'm just doing it for consistency really.

pimterry commented on lib/utils/config.ts:
```diff
@@ -0,0 +1,109 @@
+			// Device API keys are only safe for ResinOS 2.0.3+. We could somehow obtain
+			// the expected version for this config and generate one when we know it's safe,
+			// but instead for now we fall back to app keys unless the user has explicitly opted in.
+			if (deviceApiKey != null) {
```
Fixed, and renamed the methods to be a little clearer about what they're doing.

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+
+import os = require('os');
+import Promise = require('bluebird');
+import _ = require('lodash');
```
The general approach is that we put require calls inside methods _for files that have to be loaded at startup time_, not for all files everywhere. Almost all the utils files etc just use the traditional top-level `require`, it's just the CLI setup code & the actions themselves that delay requiring modules. It's safe-ish because the utils modules like this one get loaded async in the first place, only once we're running the action that needs them, so we know they're required regardless (and each of these files is similar enough that it's not worth splitting loading further).

All of this is a bit messy though, you're totally right. I'd love to do it better, but I think the main fix for this might really be going further down the bundling & packaging route, so all requires are drastically faster in the first place, and we can simplify the code completely. Might also be some webpack code splitting magic available, and I think Etcher are looking into some of that as we speak. We'll see...

thgreasi replied:
Thanks for the detailed insights :+1:

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+};
+
+export function stateToString(state: OperationState) {
+	const percentage = _.padStart(`${state.percentage}`, 3, '0') + '%';
```
Fixed

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+		case 'run-script':
+			return `${result} ${state.operation.script}`;
+		default:
+			throw new Error(`Unsupported operation: ${state.operation.command}`);
```
Indeed :cry:

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+
+	command = _.union(_.take(process.argv, 2), command);
+
+	const presidentExecuteAsync = Promise.promisify(execute);
```
Done

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+			primary: 1
+		},
+		path: '/device-type.json'
+	}).then(Promise.promisify(rindle.extract))
```
Done & done

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+	}).then(Promise.promisify(rindle.extract))
+	.then(JSON.parse)
+	.catch(() => resin.models.device.getManifestBySlug(deviceType));
+};
```
Fixed by adding linting

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+
+	step.on('state', function(state) {
+		if (state.operation.command === 'burn') { return; }
+		return console.log(exports.stateToString(state));
```
True, fixed

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+
+	step.on('burn', state => progressBars[state.type].update(state));
+
+	return Promise.promisify(rindle.wait)(step);
```
Yep, definitely seems like it. It was always the last thing running I think, so the process just stayed open until it finished anyway, but it's mostly luck that that worked :cry:

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+		getApplication(applicationName),
+		resin.models.config.getDeviceTypes(),
+		function (app, deviceTypes) {
+			let config = _.find(deviceTypes, { 'slug': app.device_type });
```
Fixed & fixed

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+		function (app, deviceTypes) {
+			let config = _.find(deviceTypes, { 'slug': app.device_type });
+
+			if (config == null) {
```
Fixed

pimterry commented on lib/utils/helpers.ts:
```diff
@@ -0,0 +1,146 @@
+};
+
+function getApplication(applicationName: string) {
+	let match;
```
Fixed

pimterry commented on lib/utils/logger.ts:
```diff
@@ -0,0 +1,66 @@
+			if (key !== 'debug') {
+				return stream.pipe(process.stdout);
+			} else {
+				if (process.env.DEBUG != null) { return stream.pipe(process.stdout); }
```
Fixed by simplifying the above

pimterry commented on lib/utils/patterns.ts:
```diff
@@ -0,0 +1,228 @@
+			});
+		});
+	}).then((application) => {
+		if (application != null) return application;
```
Fixed

pimterry commented on lib/utils/streams.ts:
```diff
@@ -0,0 +1,20 @@
+	const fileWriteStream = fs.createWriteStream(bufferFile);
+
+	return new Promise(function(resolve, reject) {
+		return stream
```
yes indeed, fixed

pimterry commented on lib/utils/update.ts:
```diff
@@ -0,0 +1,51 @@
+}
+
+export function hasAvailableUpdate() {
+	return notifier != null;
```
Yes, and it's also a weirdly named method (this isn't checking for an update, just if the update notifier was loaded). I've inlined it, and simplified en route too.

pimterry commented on tsconfig.json:
```diff
@@ -8,9 +8,20 @@
 		"preserveConstEnums": true,
 		"removeComments": true,
-		"sourceMap": true
+		"sourceMap": true,
```
As above - those are already covered by the other config here.

pimterry commented on typings/resin-device-init.d.ts:
```diff
@@ -0,0 +1,59 @@
+		on(event: 'burn', callback: (state: BurnProgress) => void): void;
+	}
+
+	export function initialize(image: string, deviceType: string, config: {}): Promise<InitializeEmitter>
```
It's not strictly necessary, but it is a good idea, done.

pimterry commented on typings/resin-image-fs.d.ts:
```diff
@@ -0,0 +1,5 @@
+declare module 'resin-image-fs' {
+	import Promise = require('bluebird');
+
+	export function read(options: {}): Promise<NodeJS.ReadableStream>;
```
Not sure what you mean? We're already importing bluebird the line just above this.

thgreasi replied:
Ohhh my, I might be getting blind, or cross posted this here instead of the other file.
Sorry for the noise, forget about this :kissing: (<= that's me whistling)

pimterry commented:
Updated to handle pretty much all the comments I think, and rebased. I've also added TS & CS linting with the new resin-lint TS support, which solved a few comments en route and should make things a little tidier generally.
